modules/sdk-lib-mpc/src/tss/ecdsa/paillierBlumProof.ts

```diff
@@ -0,0 +1,134 @@
+import { createHmac } from 'crypto';
+import { bitLength, randBits, isProbablyPrime } from 'bigint-crypto-utils';
+import { modInv, modPow } from 'bigint-mod-arith';
+import { bigIntFromBufferBE, bigIntToBufferBE } from '../../util';
+import { DeserializedPaillierBlumProof } from './types';
+
+// Security parameter.
+const m = 80;
+
+/**
+ * Generate psuedo-random quadratic residue for (N, w, i).
+ * @param N - the prime number to verify is a product of two large primes.
+ * @param w - a random number with the same bitLength as N, that satisfies the Jacobi of w is -1 wrt N.
+ * @returns {bigint[]} - set of challenges for N
+ */
+function generateY(N, w): bigint[] {
+  const NBuf = bigIntToBufferBE(N);
+  const wBuf = bigIntToBufferBE(w, NBuf.length);
+  return Array(m)
+    .fill(null)
+    .map((_, i) => {
+      const h = bigIntFromBufferBE(
+        createHmac('sha256', Buffer.from([i]))
+          .update(NBuf)
+          .update(wBuf)
+          .digest()
+      );
+      return h * h;
+    });
+}
+
+// https://en.wikipedia.org/wiki/Jacobi_symbol#Implementation_in_C++
+function jacobi(a, n): bigint {
+  // a/n is represented as (a,n)
+  if (n <= BigInt(0)) {
+    throw new Error('n must greater than 0');
+  }
+  if (n % BigInt(2) !== BigInt(1)) {
+    throw new Error('n must be odd');
+  }
+  // step 1
+  a = a % n;
+  let t = BigInt(1);
+  let r;
+  // step 3
+  while (a !== BigInt(0)) {
+    // step 2
+    while (a % BigInt(2) === BigInt(0)) {
+      a /= BigInt(2);
+      r = n % BigInt(8);
+      if (r === BigInt(3) || r === BigInt(5)) {
+        t = -t;
+      }
+    }
+    // step 4
+    r = n;
+    n = a;
+    a = r;
+    if (a % BigInt(4) === BigInt(3) && n % BigInt(4) === BigInt(3)) {
+      t = -t;
+    }
+    a = a % n;
+  }
+  if (n === BigInt(1)) {
+    return t;
+  }
+  return BigInt(0);
+}
+
+/**
+ * Prove that a modulus (p*q) is the product of two large safe primes (p and q).
+ * @param {bigint} p The larger prime factor of the modulus
+ * @param {bigint} q The smaller prime factor of the modulus.
+ * @returns {DeserializedPaillierBlumProof} The proof that the modulus is the product of two large primes.
+ */
+export async function prove(p: bigint, q: bigint): Promise<DeserializedPaillierBlumProof> {
+  // Prover selects random w with Jacobi symbol 1 wrt N.
+  const N = p * q;
+  const l = (p - BigInt(1)) * (q - BigInt(1));
+  const d = modInv(N, l);
+  let w;
+  while (true) {
+    w = bigIntFromBufferBE(Buffer.from(await randBits(bitLength(N))));
+    if (jacobi(w, N) === BigInt(-1)) {
+      break;
+    }
+  }
+  // Prover generates y_i.
+  const y = generateY(N, w);
+  // Prover calculates z_i = y_i ^ d mod N
+  const z = y.map((y_i) => modPow(y_i, d, N));
+  // Prover calculates x_i = y_i ^ 1/4 mod N using [HOC - Fact 2.160]
```

Comment:
a link would be useful here, you can reference this https://cacr.uwaterloo.ca/hac/about/chap2.pdf

Reply:
you can change this to -> 
This is calculating the inverse of the function y^4 mod N, i.e. y^(1/4), where N = pq is a blum integer using HOC - Fact 2.160 from https://cacr.uwaterloo.ca/hac/about/chap2.pdf

Reply:
```
  // This is calculating the inverse of the function y^4 mod N,
  // i.e.y ^ (1 / 4), where N = pq is a blum integer using HOC - Fact 2.160
  // from cacr.uwaterloo.ca / hac / about / chap2.pdf
  // Prover generates y_i.
  const y = generateY(N, w);
  // Prover calculates z_i = y_i ^ d mod N
  const z = y.map((y_i) => modPow(y_i, d, N));
```

like this?